include/aspect/mesh_deformation/fastscape.h

```diff
@@ -0,0 +1,680 @@
+       */
+      void execute_fastscape(double *h, double *kd, int istep) const;
+
+      /**
+       * Function to apply orographic controls to the FastScape model.
+       */
+      void apply_orographic_controls(double *h, double *kf, double *kd) const;
+
+      /**
+       * Fill velocity data table to be interpolated back onto the ASPECT mesh.
+       */
+      Table<dim,double> fill_data_table(std::vector<double> values, TableIndices<dim> size_idx, int nx, int ny) const;
+
+      /**
+       * Read data from file for restarting.
+       */
+      void read_restart_files(double *h, double *b, double *sf) const;
+
+      /**
+       * Save data to file for restarting.
+       */
+      void save_restart_files(const double *h, double *b, double *sf, int istep) const;
+
+    private:
+      /**
+       * First attempt at the number of steps to run FastScape for every ASPECT timestep,
+       * where the FastScape timestep is determined by (ASPECT_timestep/nstep).
+       */
+      int nstep;
+
+      /**
+       * Maximum timestep allowed for FastScape, if (ASPECT_timestep/nsteps) exceeds this, nsteps is doubled.
+       */
+      double maximum_fastscape_timestep;
+
+      /**
+       * Check whether FastScape needs to be restarted. This is used as
+       * a mutable bool because we determine whether the model is being resumed in
+       * initialize(), and then after reinitializing FastScape we change it to zero
+       * so it does not initialize FastScape again in future timesteps.
+       * TODO: There is probably a better way to do this, and restarts should be rolled into
+       * the general ASPECT restart.
+       */
+      mutable bool restart;
+
+      /**
+       * FastScape resets to step 0 on restart, this keeps the step number from the previous run.
+       * As we only want to read this value in from the restart file once, we initialize it in
+       * initialize() and only change it during the first timestep where FastScape is initialized.
+       */
+      mutable int restart_step;
+
+      /**
+       * ASPECT end time to check if we should destroy FastScape.
+       */
+      double end_time;
+
+      /**
+       * FastScape cell size in X, dx should always equal dy.
+       */
+      double dx;
+
+      /**
+       * FastScape cell size in Y, dy should always equal dx.
+       */
+      double dy;
+
+      /**
+       * FastScape X extent (ASPECT X extent + 2*dx for ghost nodes).
+       */
+      double x_extent;
+
+      /**
+       * Fastscape Y extent (ASPECT Y extent + 2*dy for ghost nodes).
+       */
+      double y_extent;
+
+      /**
+       * User set FastScape Y extent for a 2D ASPECT model.
+       */
+      double y_extent_2d;
+
+      /**
+       * Number of x points in FastScape array.
+       */
+      int nx;
+
+      /**
+       * Number of y points in FastScape array.
+       */
+      int ny;
+
+      /**
+       * Size of the FastScape array (nx*ny).
+       */
+      int array_size;
```

Comment:
can this be an `unsigned int`? That would be more precise. Same question for nx and ny above.

Reply:
I thought I had checked this previously, but array_size isn't called by fastscape so this can be an unsigned int. Ny and nx, however, cannot be unsigned. I've updated all the array_size loops to reflect this.